### DISCUSSION TITLE
Getting Huskylens working on SPIKE Prime with Pybricks (All credit to Daniele Benedettelli’s & Anton’s Mindstorms' work)

### DISCUSSION THREAD

**Discussion by JDwyer009** [pybricks/support#1174] [General]
Hi All,

With the transition from SPIKE App v2 to SPIKE App v3 a lot of the functionality previously available in the Python Editor has been removed.   
I’m trying to convert a few projects that used to work in SPIKE App v2 to Pybricks. Obviously, I’m having trouble. I’m close to my current limits of understanding and would love some help.

In this case I’m trying to use **Daniele Benedettelli’s & Anton’s Mindstorms** really cool projects getting the Huskylens camera to work with a SPIKE Prime, to now work on Pybricks. 

The major issues I can identify, is that previously there was a function in the Port Class that allowed you to set the mode (duplex), baud rate and control pwm on the SPIKE Hub Ports.

Now what I get is errors such as: "AttributeError: ‘Port’ object has no attribute ‘mode’"

The answers appear to be in [here,](https://github.com/pybricks/technical-info/blob/master/uart-protocol.md) but at this stage coding at that level is still beyond me.

Any suggestions?  

Briefly, the most recent errors I’m experiencing start happening at line 76, 78 & 80. 
    self.uart.mode(1)
    self.uart.baud(baudrate)
    self.uart.pwm(100)
See in context below + I’ve added a partially transitioned Beneddettelli file.
```
class HuskyLensCamera:
    def __init__(self, port = 0, baudrate = 115200, debug=True):
        self.checkOnceAgain=True
        self.DEBUG=debug
        self.unprocessed_data=b''
        self.baudrate = baudrate
        self.reads_per_ms = 10
        if type(port) == str:
            self.uart = eval("hub.port."+port)
        else:
            self.uart = port
        self.uart.mode(1)
        wait(300)# wait for all duplex methods to appear
        self.uart.baud(baudrate) # set baud rate
        wait(200)# wait for all duplex methods to appear
        self.uart.pwm(100)
        wait(2200)
        self.lastCmdSent = ""
        self.flush()
        self.knock()
```

[PyBricks_HuskyLens.zip](https://github.com/pybricks/support/files/12142297/PyBricks_HuskyLens.zip)

**Comment by laurensvalk** [comment]
Pybricks does not currently support direct control of uart signals. See https://github.com/pybricks/support/issues/220.

If we can find enough users to rally around the cause, maybe it will happen some day. If everyone who purchased custom hardware would sponsor Pybricks by the same amount, that would be even better. :smile:

**Reply by JDwyer009** [reply]
Thanks for the quick response. 
Obviously it was a futile endeavor, from the start. :)

That's a bummer, if Pybricks doesn't allow 3rd party UART devices and the latest LEGO SPIKE App no longer supports it and neither have any plans to implement it.

**Reply by laurensvalk** [reply]
It's not that we don't allow it, we just haven't gotten around to supporting it so far.

It does sound like a fun project!

**Reply by JDwyer009** [reply]
Pybricks is clearly testament to the fact you enjoy a challenge!! And have the skills!
Clearly there has been some thought put into it. 
Just saying. :)

**Reply by meimei-code** [reply]
@JDwyer009  May I kindly inquire if you have managed to resolve the aforementioned issue?

**Comment by meimei-code** [comment]
@laurensvalk  We intend to create a facial recognition project using the Huskylens AI vision sensor and Spike Prime. However, we have encountered an issue with the firmware and LEGO Education Spike software of the Spike Prime controller, as they have been updated to version 3.0 and no longer support the "Pybricks" library (which was integrated into the version 2.0 firmware). This means that Spike Prime cannot drive the external vision sensor, Huskylens. Have you come across this problem? Is there a plan to update and upgrade the Pybricks library to establish compatibility between Spike Prime version 3.0 and Huskylens? Alternatively, are there any other solutions available?

**Reply by laurensvalk** [reply]
I don't think the Pybricks library was included in the LEGO SPIKE 2.0 version.

If you'd like to see support for the Huskylens, please consider asking the manufacturer to reach out to us at team@pybricks.com.

**Reply by weichunmei** [reply]
I want to complete the project detailed in the post '[How to use the Huskylens with LEGO SPIKE Prime or Robot Inventor](https://www.antonsmindstorms.com/2021/10/17/how-to-use-the-huskylens-with-lego-spike-prime-or-robot-inventor/)', but I'm having trouble using the Huskylens. There's an error when I'm compiling the program. It seems that Spike Prime V3.0 does not support the 'Pybricks' library.

**Reply by JDwyer009** [reply]
@meimei-code & @weichunmei 

Apologies for the late reply to above. In short, I have not been able to resolve the above issue.

To the best of my knowledge, at present, the only SPIKE Prime programming platform that supports the Huskylens project is the SPIKE App 2.0 (aka SPIKE Legacy App). Neither SPIKE App 3.0 or Pybricks support the necessary UART protocol. 

However, we can always hold out hope than when the Pybricks team get their current major project completed, support for the UART protocol will sit on the list of 'to do' tasks.
